### PR TITLE
Track frequency of unhandled JS errors with MC Stats

### LINKED ISSUE #49289: [Remote Error Logging] Use MC Stats for Error Counting
## Description

Implement MC Stats for error counting to monitor the number of fatal errors that occur per day, week, and month. This needs to be done on both the JS and PHP sides.

The implementation should be similar to existing Tracks tacking.

## Reference

- Calypso [`bumpStat()`](https://github.com/Automattic/wp-calypso/blob/31d3844d216c261ed9bff5ea0abadc0aa0d78b4b/client/lib/analytics/mc.js#L34-L50)
- PCYsg-J0-p2
- PCYsg-G3-p2

## Accept Criteria

1. Only count errors when users have opted in to tracking/logging.
2. Implement MC Stats for fatal error counting on the JS side.
3. Implement MC Stats for fatal error counting on the PHP side.
4. Develop unit tests for the MC Stats implementation on both JS and PHP sides.

### PR DESCRIPTION
**Pull request by chihsuan**
Track frequency of unhandled JS errors with MC Stats

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes https://github.com/woocommerce/woocommerce/issues/49289

This PR adds a new function, `bumpStat,` to the `@woocommerce/tracks` package and uses it to track the frequency of unhandled JS errors.

The implementation is based on Calypso's [`bumpStat()`](https://github.com/Automattic/wp-calypso/blob/31d3844d216c261ed9bff5ea0abadc0aa0d78b4b/client/lib/analytics/mc.js#L34-L50). 

Reference: PCYsg-J0-p2

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

To test this PR, we need to use the local dev environment and follow the below steps to simulate the error and check if it is logged in the server.

1. Checkout this PR branch.
2. Run `pnpm install`
3.  Install Code Snippets plugin and add a new snippet with the below code to enable the remote logging.

		```php
		update_option( 'woocommerce_feature_remote_logging_enabled', 'yes' );
		update_option( 'woocommerce_allow_tracking', 'yes' );

		add_filter(
			'option_woocommerce_remote_variant_assignment',
			function () {
				return 5;
			}
		);

4. Go to `WooCommerce > Home` page.
5. Open the browser dev tools and go to the console tab.
6. Run `window.localStorage.setItem( 'debug', '*')` to enable the debug logging.
7. Now, run `git cherry-pick 146a650` (146a650) to simulate the error.
8. Run `pnpm --filter='@woocommerce/plugin-woocommerce' build` to build production files.
9. Confirm `wc-admin:tracks:stats Bumping stat error:unhandled-js-errors` is logged in the console.
10. Reload the page and search `http://pixel.wp.com/g.gif?v=wpcom-no-pv&x_woocommerce-error` in the network tab to confirm the stats are sent to the server. (It would take hours to days to see the stats in the server)
11. Run `git cherry-pick 701c795bf16874bd00b4739e871ef1da945ffa83` (701c795bf16874bd00b4739e871ef1da945ffa83) to simulate the error while rendering the component.
12. Run `pnpm --filter='@woocommerce/plugin-woocommerce' build`
13. Reload the page and search `http://pixel.wp.com/g.gif?v=wpcom-no-pv&x_woocommerce-error` in the network tab to confirm the stats are sent to the server.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

### CODE REVIEW

**Review by psealock — APPROVED** [review]
This is also looking good and is testing well. I can see the correct payload applied to the pixel request.

File: packages/js/tracks/src/stats.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+ *
+ * @param {Record<string, string> | string} group  - The group of stats or a single stat name.
+ * @param {string}                          [name] - The name of the stat if group is a string.
+ * @return {string} The constructed querystring.
+ */
+function buildQuerystring(
+	group: Record< string, string > | string,
+	name?: string
+): string {
+	let uriComponent = '';
```
Comment (rjchow) [inline packages/js/tracks/src/stats.ts]:
```suggestion
const PREFIX = 'x_woocommerce-'
 if (typeof group !== object) { return (new URLSearchParams( { `${PREFIX}${group}` : name } )).toString();
 else {
  const prefixedGroups = {};
  Object.entries(group).forEach(([key, value]) => { 
    prefixedGroups[`${PREFIX}${key}`] = value;
   })
   return (new URLSearchParams(prefixedGroups)).toString();
 }

```

I wrote this in the comment box so it's not 100% tested but I think this should work!

File: packages/js/tracks/src/stats.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+		!! window.wcTracks &&
+		!! window.wcTracks.isEnabled;
+
+	if ( ! shouldBumpStat ) {
+		return false;
+	}
+
+	const uriComponent = buildQuerystring( group, name );
+	new window.Image().src =
+		document.location.protocol +
```
Comment (rjchow) [inline packages/js/tracks/src/stats.ts]: Like above I think using URLSearchParams() would make this safer

**Review by rjchow — COMMENTED** [review]
Code-only review since its pretty straightforward

File: packages/js/tracks/src/stats.ts (hunk 1 above)
Comment (chihsuan) [reply packages/js/tracks/src/stats.ts]: This is a great suggestion! @rjchow  I've made the changes based on your feedback. 🙌  f858ff5ed095b4f72b722b6af5ca67e393c3c2d0

**Review by rjchow — APPROVED** [review]
Thanks! That looks even better than what I had in mind 😄

**Pull request merged**